File: app/src/main/java/com/dreams/androidquizapp/fragments/QuestionFragment.java
package com.dreams.androidquizapp.fragments;

import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.dreams.androidquizapp.MainActivity;
import com.dreams.androidquizapp.OnFragmentInteractionListener;
import com.dreams.androidquizapp.R;
import com.dreams.androidquizapp.models.Answer;
import com.dreams.androidquizapp.models.Question;

import java.util.ArrayList;
import java.util.Random;


/**
 * A simple {@link Fragment} subclass.
 * Activities that contain this fragment must implement the
 * {@link OnFragmentInteractionListener} interface
 * to handle interaction events.
 * Use the {@link QuestionFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class QuestionFragment extends Fragment implements View.OnClickListener {

	// Declare Constants
	private static final String QUESTION = "Question";
	private static final String KEY = "Key";
	private static final String DETAILS = "Details";
	private static final String FILLANSWERS = "Fill";

	// Declare UI Variables
	private TextView questionTV;
	private TextView detailsA, detailsB, detailsC, detailsD;
	private TextView answerA, answerB, answerC, answerD;

	// Declare Variables
	private String quizQuestion;
	private String quizAnswer, quizDetails;
	private ArrayList<Answer> otherAnswers;
	private Button nextBtn, newBtn;
	private boolean[] answerCheck = new boolean[4];
	private OnFragmentInteractionListener mListener;
	private int keyPosition;
	private boolean answered = false;
	private boolean correct = false;

	public QuestionFragment() {
		// Required empty public constructor
	}

	/**
	 * Use this factory method to create a new instance of
	 * this fragment using the provided parameters.
	 *
	 * @param question Accepts a Question object.
	 * @param fillAnswers Accepts an array of Incorrect Answer objects
	 * @return A new instance of fragment QuestionFragment.
	 */
	public static QuestionFragment newInstance(Question question, ArrayList<Answer> fillAnswers) {
		// Create new instance of fragment
		QuestionFragment fragment = new QuestionFragment();
		// Add arguments to bundle of new instance.
		Bundle args = new Bundle();
		args.putString(QUESTION, question.getQuestion());
		args.putString(KEY, question.getShortAns());
		args.putString(DETAILS, question.getDetails());
		args.putParcelableArrayList(FILLANSWERS, fillAnswers);

		fragment.setArguments(args);
		return fragment;
	}


	@Override
	public void onAttach(Context context) {
		super.onAttach(context);
		if (context instanceof OnFragmentInteractionListener) {
			mListener = (OnFragmentInteractionListener) context;
		} else {
			throw new RuntimeException(context.toString()
							+ " must implement OnFragmentInteractionListener");
		}
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		// Get Arguments from bundle
		if (getArguments() != null) {
			quizQuestion = getArguments().getString(QUESTION);
			quizAnswer = getArguments().getString(KEY);
			quizDetails = getArguments().getString(DETAILS);
			otherAnswers = getArguments().getParcelableArrayList(FILLANSWERS);


		}
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
													 Bundle savedInstanceState) {
		// Inflate the layout for this fragment
		View mView = inflater.inflate(R.layout.fragment_question, container, false);

		// Assign Variables to Object Ids
		questionTV = (TextView) mView.findViewById(R.id.question);
		answerA = (TextView) mView.findViewById(R.id.answer_a_summary);
		detailsA = (TextView) mView.findViewById(R.id.answer_a_details);
		answerB = (TextView) mView.findViewById(R.id.answer_b_summary);
		detailsB = (TextView) mView.findViewById(R.id.answer_b_details);
		answerC = (TextView) mView.findViewById(R.id.answer_c_summary);
		detailsC = (TextView) mView.findViewById(R.id.answer_c_details);
		answerD = (TextView) mView.findViewById(R.id.answer_d_summary);
		detailsD = (TextView) mView.findViewById(R.id.answer_d_details);
		nextBtn = (Button) mView.findViewById(R.id.nextBtn);
		newBtn = (Button) mView.findViewById(R.id.newQuizBtn);

		//Set Listeners
		nextBtn.setOnClickListener(this);
		newBtn.setOnClickListener(this);
		answerA.setOnClickListener(this);
		answerB.setOnClickListener(this);
		answerC.setOnClickListener(this);
		answerD.setOnClickListener(this);


		// Set the Question Text
		questionTV.setText(quizQuestion);

		// Create random and int for key positioning.
		Random random = new Random(System.currentTimeMillis());
		keyPosition = random.nextInt(4);

		// Position Key according to random int.
		switch (keyPosition) {
			case 0:
				answerA.setText(quizAnswer);
				detailsA.setText(quizDetails);
				answerCheck[0] = true;
				break;
			case 1:
				answerB.setText(quizAnswer);
				detailsB.setText(quizDetails);
				answerCheck[1] = true;
				break;
			case 2:
				answerC.setText(quizAnswer);
				detailsC.setText(quizDetails);
				answerCheck[2] = true;
				break;
			case 3:
				answerD.setText(quizAnswer);
				detailsD.setText(quizDetails);
				answerCheck[3] = true;
				break;
		}
		int randomAnswer;
		for (int j = 0; j < answerCheck.length; j++) {
			// if position in answer check =/= true, fill with random answer
			if (!answerCheck[j]) {
				randomAnswer = random.nextInt(otherAnswers.size());
				switch (j) {
					case 0:
						answerA.setText((String.valueOf(otherAnswers.get(randomAnswer).getAnswer())));
						detailsA.setText((String.valueOf(otherAnswers.get(randomAnswer).getDetails())));
						answerCheck[0] = true;
						break;
					case 1:
						answerB.setText((String.valueOf(otherAnswers.get(randomAnswer).getAnswer())));
						detailsB.setText((String.valueOf(otherAnswers.get(randomAnswer).getDetails())));
						answerCheck[1] = true;
						break;
					case 2:
						answerC.setText((String.valueOf(otherAnswers.get(randomAnswer).getAnswer())));
						detailsC.setText((String.valueOf(otherAnswers.get(randomAnswer).getDetails())));
						answerCheck[2] = true;
						break;
					case 3:
						answerD.setText((String.valueOf(otherAnswers.get(randomAnswer).getAnswer())));
						detailsD.setText((String.valueOf(otherAnswers.get(randomAnswer).getDetails())));
						answerCheck[3] = true;
						break;
				}
			}
		}

		return mView;
	}

	@Override
	public void onDetach() {
		super.onDetach();
		mListener = null;
	}

	@Override
	public void onClick(View view) {

		// Switch statement for what is clicked.
		switch (view.getId()) {
			case R.id.nextBtn:
				((MainActivity) getActivity()).nextQuestion(correct);
				break;
			case R.id.newQuizBtn:
				// start main Activity over
				Intent intent = new Intent(getActivity(), MainActivity.class);
				startActivity(intent);
				// remove previous from backstack
				getActivity().finish();
				break;
			case R.id.answer_a_summary:
				if (detailsA.getVisibility() != View.VISIBLE) {
					detailsA.setVisibility(View.VISIBLE);

					// Check if question has not been answered
					if (!answered) {

						// Check if this selection is correct answer
						if (keyPosition == 0) {
							// Right answer is A
							detailsA.setTextColor(Color.GREEN);
							answerA.setTextColor(Color.GREEN);

							// Register the correct answer
							correct = true;
						}else{
							// Incorrect Answer Selected
							detailsA.setTextColor(Color.RED);
							answerA.setTextColor(Color.RED);
						}
						answered = true;
					}
				}
				break;
			case R.id.answer_b_summary:
				if (detailsB.getVisibility() != View.VISIBLE) {
					detailsB.setVisibility(View.VISIBLE);

					// Check if question has not been answered
					if (!answered) {

						// Check if this selection is correct answer
						if (keyPosition == 1) {
							// Right answer is A
							detailsB.setTextColor(Color.GREEN);
							answerB.setTextColor(Color.GREEN);

							// Register the correct answer
							correct = true;
						}else{
							// Incorrect Answer Selected
							detailsB.setTextColor(Color.RED);
							answerB.setTextColor(Color.RED);
						}
						answered = true;
					}
				}
				break;
			case R.id.answer_c_summary:
				if (detailsC.getVisibility() != View.VISIBLE) {
					detailsC.setVisibility(View.VISIBLE);

					// Check if question has not been answered
					if (!answered) {

						// Check if this selection is correct answer
						if (keyPosition == 2) {
							// Right answer is A
							detailsC.setTextColor(Color.GREEN);
							answerC.setTextColor(Color.GREEN);

							// Register the correct answer
							correct = true;
						}else{
							// Incorrect Answer Selected
							detailsC.setTextColor(Color.RED);
							answerC.setTextColor(Color.RED);
						}
						answered = true;
					}
				}
				break;
			case R.id.answer_d_summary:
				if (detailsD.getVisibility() != View.VISIBLE) {
					detailsD.setVisibility(View.VISIBLE);
					// Check if question has not been answered
					if (!answered) {
						// Check if this selection is correct answer
						if (keyPosition == 3) {
							// Right answer is A
							detailsD.setTextColor(Color.GREEN);
							answerD.setTextColor(Color.GREEN);
							// Register the correct answer
							correct = true;
						}else{
							// Incorrect Answer Selected
							detailsD.setTextColor(Color.RED);
							answerD.setTextColor(Color.RED);
						}
						answered = true;
					}
				}
				break;
		}
	}
}

File: Interview_Prep_App/native_kotlin/retrofit/src/main/java/com/interviewprep/kotlinretrofit/fragments/QuestionFragment.kt
package com.interviewprep.kotlinretrofit.fragments

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.interviewprep.kotlinretrofit.MainActivity
import com.interviewprep.kotlinretrofit.OnFragmentInteractionListener
import com.interviewprep.kotlinretrofit.R
import com.interviewprep.kotlinretrofit.models.Answer
import com.interviewprep.kotlinretrofit.models.Question
import java.util.Random

/**
 * A simple [Fragment] subclass.
 * Activities that contain this fragment must implement the
 * [OnFragmentInteractionListener] interface
 * to handle interaction events.
 * Use the [QuestionFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
class QuestionFragment : Fragment(), View.OnClickListener {
    // Declare UI Variables
    private var questionTV: TextView? = null
    private var detailsA: TextView? = null
    private var detailsB: TextView? = null
    private var detailsC: TextView? = null
    private var detailsD: TextView? = null
    private var answerA: TextView? = null
    private var answerB: TextView? = null
    private var answerC: TextView? = null
    private var answerD: TextView? = null

    // Declare Variables
    private var quizQuestion: String? = null
    private var quizAnswer: String? = null
    private var quizDetails: String? = null
    private var otherAnswers: ArrayList<Answer>? = null
    private var nextBtn: Button? = null
    private var newBtn: Button? = null
    private val answerCheck = BooleanArray(4)
    private var mListener: OnFragmentInteractionListener? = null
    private var keyPosition = 0
    private var answered = false
    private var correct = false
    override fun onAttach(context: Context) {
        super.onAttach(context)
        mListener = if (context is OnFragmentInteractionListener) {
            context
        } else {
            throw RuntimeException(context.toString() + " must implement OnFragmentInteractionListener")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // Get Arguments from bundle
        if (arguments != null) {
            quizQuestion = requireArguments().getString(QUESTION)
            quizAnswer = requireArguments().getString(KEY)
            quizDetails = requireArguments().getString(DETAILS)
            otherAnswers = requireArguments().getParcelableArrayList(FILLANSWERS)
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        val mView = inflater.inflate(R.layout.fragment_question, container, false)

        // Assign Variables to Object Ids
        questionTV = mView.findViewById(R.id.question) as TextView
        answerA = mView.findViewById(R.id.answer_a_summary) as TextView
        detailsA = mView.findViewById(R.id.answer_a_details) as TextView
        answerB = mView.findViewById(R.id.answer_b_summary) as TextView
        detailsB = mView.findViewById(R.id.answer_b_details) as TextView
        answerC = mView.findViewById(R.id.answer_c_summary) as TextView
        detailsC = mView.findViewById(R.id.answer_c_details) as TextView
        answerD = mView.findViewById(R.id.answer_d_summary) as TextView
        detailsD = mView.findViewById(R.id.answer_d_details) as TextView
        nextBtn = mView.findViewById(R.id.nextBtn) as Button
        newBtn = mView.findViewById(R.id.newQuizBtn) as Button

        // Set Listeners
        nextBtn!!.setOnClickListener(this)
        newBtn!!.setOnClickListener(this)
        answerA!!.setOnClickListener(this)
        answerB!!.setOnClickListener(this)
        answerC!!.setOnClickListener(this)
        answerD!!.setOnClickListener(this)

        // Set the Question Text
        questionTV!!.text = quizQuestion

        // Create random and int for key positioning.
        val random = Random(System.currentTimeMillis())
        keyPosition = random.nextInt(4)
        when (keyPosition) {
            0 -> {
                answerA!!.text = quizAnswer
                detailsA!!.text = quizDetails
                answerCheck[0] = true
            }
            1 -> {
                answerB!!.text = quizAnswer
                detailsB!!.text = quizDetails
                answerCheck[1] = true
            }
            2 -> {
                answerC!!.text = quizAnswer
                detailsC!!.text = quizDetails
                answerCheck[2] = true
            }
            3 -> {
                answerD!!.text = quizAnswer
                detailsD!!.text = quizDetails
                answerCheck[3] = true
            }
        }
        var randomAnswer: Int
        for (j in answerCheck.indices) {
            // if position in answer check =/= true, fill with random answer
            if (!answerCheck[j]) {
                randomAnswer = random.nextInt(otherAnswers!!.size)
                when (j) {
                    0 -> {
                        answerA!!.text = otherAnswers!![randomAnswer].answer.toString()
                        detailsA!!.text = otherAnswers!![randomAnswer].details.toString()
                        answerCheck[0] = true
                    }
                    1 -> {
                        answerB!!.text = otherAnswers!![randomAnswer].answer.toString()
                        detailsB!!.text = otherAnswers!![randomAnswer].details.toString()
                        answerCheck[1] = true
                    }
                    2 -> {
                        answerC!!.text = otherAnswers!![randomAnswer].answer.toString()
                        detailsC!!.text = otherAnswers!![randomAnswer].details.toString()
                        answerCheck[2] = true
                    }
                    3 -> {
                        answerD!!.text = otherAnswers!![randomAnswer].answer.toString()
                        detailsD!!.text = otherAnswers!![randomAnswer].details.toString()
                        answerCheck[3] = true
                    }
                }
            }
        }
        return mView
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    override fun onClick(view: View) {
        // Switch statement for what is clicked.
        when (view.id) {
            R.id.nextBtn -> (activity as MainActivity).nextQuestion(correct)
            R.id.newQuizBtn -> {
                // start main Activity over
                val intent = Intent(activity, MainActivity::class.java)
                startActivity(intent)
                // remove previous from backstack
                activity?.finish()
            }
            R.id.answer_a_summary -> if (detailsA!!.visibility != View.VISIBLE) {
                detailsA!!.visibility = View.VISIBLE

                // Check if question has not been answered
                if (!answered) {
                    // Check if this selection is correct answer
                    if (keyPosition == 0) {
                        // Right answer is A
                        detailsA!!.setTextColor(Color.GREEN)
                        answerA!!.setTextColor(Color.GREEN)

                        // Register the correct answer
                        correct = true
                    } else {
                        // Incorrect Answer Selected
                        detailsA!!.setTextColor(Color.RED)
                        answerA!!.setTextColor(Color.RED)
                    }
                    answered = true
                }
            }
            R.id.answer_b_summary -> if (detailsB!!.visibility != View.VISIBLE) {
                detailsB!!.visibility = View.VISIBLE

                // Check if question has not been answered
                if (!answered) {
                    // Check if this selection is correct answer
                    if (keyPosition == 1) {
                        // Right answer is A
                        detailsB!!.setTextColor(Color.GREEN)
                        answerB!!.setTextColor(Color.GREEN)

                        // Register the correct answer
                        correct = true
                    } else {
                        // Incorrect Answer Selected
                        detailsB!!.setTextColor(Color.RED)
                        answerB!!.setTextColor(Color.RED)
                    }
                    answered = true
                }
            }
            R.id.answer_c_summary -> if (detailsC!!.visibility != View.VISIBLE) {
                detailsC!!.visibility = View.VISIBLE

                // Check if question has not been answered
                if (!answered) {
                    // Check if this selection is correct answer
                    if (keyPosition == 2) {
                        // Right answer is A
                        detailsC!!.setTextColor(Color.GREEN)
                        answerC!!.setTextColor(Color.GREEN)

                        // Register the correct answer
                        correct = true
                    } else {
                        // Incorrect Answer Selected
                        detailsC!!.setTextColor(Color.RED)
                        answerC!!.setTextColor(Color.RED)
                    }
                    answered = true
                }
            }
            R.id.answer_d_summary -> if (detailsD!!.visibility != View.VISIBLE) {
                detailsD!!.visibility = View.VISIBLE
                // Check if question has not been answered
                if (!answered) {
                    // Check if this selection is correct answer
                    if (keyPosition == 3) {
                        // Right answer is A
                        detailsD!!.setTextColor(Color.GREEN)
                        answerD!!.setTextColor(Color.GREEN)
                        // Register the correct answer
                        correct = true
                    } else {
                        // Incorrect Answer Selected
                        detailsD!!.setTextColor(Color.RED)
                        answerD!!.setTextColor(Color.RED)
                    }
                    answered = true
                }
            }
        }
    }

    companion object {
        // Declare Constants
        private const val QUESTION = "Question"
        private const val KEY = "Key"
        private const val DETAILS = "Details"
        private const val FILLANSWERS = "Fill"

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @param question Accepts a Question object.
         * @param fillAnswers Accepts an array of Incorrect Answer objects
         * @return A new instance of fragment QuestionFragment.
         */
        fun newInstance(question: Question, fillAnswers: ArrayList<Answer>): QuestionFragment {
            // Create new instance of fragment
            val fragment = QuestionFragment()
            // Add arguments to bundle of new instance.
            val args = Bundle()
            args.putString(QUESTION, question.question)
            args.putString(KEY, question.shortAns)
            args.putString(DETAILS, question.details)
            args.putParcelableArrayList(FILLANSWERS, fillAnswers)
            fragment.arguments = args
            return fragment
        }
    }
}
